### ISSUE TITLE
bufio in readPacket causes reads to never complete from Windows AD server

### ISSUE THREAD

**Issue by JJCinAZ** [go-asn1-ber/asn1-ber#33]
After to the insertion of the bufio.NewReader() in the readPacket() function in ber.go, reads never complete from an Active Directory server.  It appears that eventually the LDAP server finally closes the connection causing the read to finally fail and return.  Otherwise, the call just appears to hang until that time.

**Comment by johnweldon** [comment]
Can you provide some sample code?  The `bufio.NewReader()` change is in version v1.5.2, but the version referenced from go-ldap is v1.5.1.  Thanks!

**Comment by johnweldon** [comment]
I was able to reproduce this against OpenLDAP - workaround is to set your go.mod version of asn1-ber to v1.5.1 explicitly.

**Comment by johnweldon** [comment]
Resolved in v1.5.3, I reverted #31 - we'll need better analysis of the buffering before including it.

**Comment by bimmlerd** [comment]
Well that's unfortunate, apologies for not catching that when proposing the change - @johnweldon do you mind sharing the reproduction so that I can maybe have a look?

**Comment by johnweldon** [comment]
I know, I should have caught this too.  The code I used was a simple search on organization objects, with that buffering it hangs for me, and without it doesn't.


```go
package main

import (
	"fmt"
	"os"

	"github.com/go-ldap/ldap/v3"
)

func main() {
	ldapURL := os.Getenv("LDAP_URL")
	ldapUser := os.Getenv("LDAP_USER")
	ldapPass := os.Getenv("LDAP_PASS")
	ldapBase := os.Getenv("LDAP_BASE")

	l, err := ldap.DialURL(ldapURL)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error dialing: %v", err)
		os.Exit(1)
	}
	defer l.Close()

	if err = l.Bind(ldapUser, ldapPass); err != nil {
		fmt.Fprintf(os.Stderr, "Error binding: %v", err)
		os.Exit(2)
	}

	req := ldap.NewSearchRequest(
		ldapBase,
		ldap.ScopeWholeSubtree,
		ldap.NeverDerefAliases,
		0, 0, false,
		"(&(objectClass=organizationalPerson))",
		[]string{"cn", "dn"},
		nil,
	)

	res, err := l.SearchWithPaging(req, 10)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error searching: %v", err)
		os.Exit(3)
	}

	for _, entry := range res.Entries {
		fmt.Fprintf(os.Stdout, "%40s %s\n", entry.GetAttributeValue("cn"), entry.DN)
	}
}
```

**Comment by bimmlerd** [comment]
TL;DR The buffering lost bytes, and we can't buffer at this layer without changing the API. We should probably buffer in `go-ldap` instead.

It took some doing, but I figured it out - and of course it's rather obvious in hindsight. What I missed is that `readPackets` (implicit) contract is that it consumes only exactly as many bytes as it has to. Introducing a buffered reader breaks this contract, since the underlying reader may read more to then buffer it. But, unfortunately, this created a race condition - if the connection had many bytes available, the buffered reader would consume as much as possible, and thus possibly more than readPacket should consume. Since the buffer is discarded when `readPacket` is called anew (i.e. not recursively) we lose the bytes which were buffered, leading to the hangs observed (since the connection stays open, we also don't read an EOF). 

I think the solution is that we can't buffer at this stage, but should introduce a `bufio.Reader` over the naked `net.*Conn` in `go-ldap`.
